src/core/AutomatableModel.cpp

```diff
@@ -458,78 +428,102 @@ float AutomatableModel::fittedValue( float value ) const
 
 
 
-void AutomatableModel::linkModel( AutomatableModel* model )
+void AutomatableModel::linkModel(AutomatableModel* model)
 {
-	auto containsModel = std::find(m_linkedModels.begin(), m_linkedModels.end(), model) != m_linkedModels.end();
-	if (!containsModel && model != this)
-	{
-		m_linkedModels.push_back( model );
+	if (model == nullptr || model == this || isModelLinked(model)) { return; }
 
-		if( !model->hasLinkedModels() )
-		{
-			QObject::connect( this, SIGNAL(dataChanged()),
-					model, SIGNAL(dataChanged()), Qt::DirectConnection );
-		}
+	AutomatableModel* otherEnd = model->getLastLinkedModel();
```

Comment:
If `getLastLinkedModel()` returned `this` rather than `nullptr` when it doesn't have a linked model, this function would be simpler since you wouldn't need the `otherEnd == nullptr` branch below.